File: src/checker/checker.ts
import type { ComponentNode, InterfaceNode, ProgramNode, PropertyNode, Scalar } from "../parser/ast.js";
export interface TamlDiagnostic { code: string; component: string; line: number; message: string; formatted: string; }
export interface CheckerResult { valid: boolean; diagnostics: TamlDiagnostic[]; components: ComponentNode[]; }
export class TamlTypeChecker {
  check(program: ProgramNode): CheckerResult {
    const interfaces = new Map<string, InterfaceNode>(); const declared = new Map<string, ComponentNode>(); const diagnostics: TamlDiagnostic[] = [];
    for (const node of program.body) { if (node.type === "Interface") { if (interfaces.has(node.name)) diagnostics.push(this.diag(node.name, node.location.line, "DUPLICATE_INTERFACE", `duplicate interface '${node.name}'`)); interfaces.set(node.name, node); } else if (node.type === "Component") { if (declared.has(node.name)) diagnostics.push(this.diag(node.name, node.line, "DUPLICATE_COMPONENT", `duplicate component '${node.name}'`)); declared.set(node.name, node); } }
    const resolved = new Map<string, ComponentNode>(); const resolving = new Set<string>();
    const resolve = (node: ComponentNode): ComponentNode => {
      const cached = resolved.get(node.name); if (cached) return cached;
      if (resolving.has(node.name)) { diagnostics.push(this.diag(node.name, node.line, "CYCLIC_INHERITANCE", "cyclic component inheritance")); return node; }
      resolving.add(node.name); let inherited: Record<string, Scalar> = {};
      let inheritedLocations: ComponentNode["valueLocations"] = {};
      if (node.extendsComponent) { const parent = declared.get(node.extendsComponent); if (!parent) diagnostics.push(this.diag(node.name, node.line, "UNKNOWN_PARENT", `extends unknown component '${node.extendsComponent}'`)); else { const resolvedParent = resolve(parent); inherited = { ...resolvedParent.values }; inheritedLocations = { ...resolvedParent.valueLocations }; } }
      const component: ComponentNode = { ...node, values: { ...inherited, ...node.values }, valueLocations: { ...inheritedLocations, ...node.valueLocations } };
      if (component.implementsInterface) { const iface = interfaces.get(component.implementsInterface); if (!iface) diagnostics.push(this.diag(node.name, node.line, "UNKNOWN_INTERFACE", `implements unknown interface '${component.implementsInterface}'`)); else this.validate(component, iface, diagnostics); }
      resolving.delete(node.name); resolved.set(node.name, component); return component;
    };
    const components = [...declared.values()].map(resolve); return { valid: diagnostics.length === 0, diagnostics, components };
  }
  private validate(component: ComponentNode, iface: InterfaceNode, diagnostics: TamlDiagnostic[]): void {
    for (const property of Object.values(iface.properties)) {
      if (!(property.key in component.values)) { if (property.defaultValue !== undefined) { component.values[property.key] = property.defaultValue; component.valueLocations[property.key] = property.location; } else if (!property.optional) diagnostics.push(this.diag(component.name, component.line, "MISSING_PROPERTY", `Property '${property.key}' is required by interface '${iface.name}'.`)); continue; }
      const value = component.values[property.key]; const issue = this.valueError(property, value); if (issue) diagnostics.push(this.diag(component.name, component.line, "TYPE_ERROR", `Property '${property.key}' value (${JSON.stringify(value)}) ${issue}.`));
    }
  }
  private valueError(property: PropertyNode, value: Scalar): string | undefined {
    const numeric = typeof value === "number" && Number.isFinite(value);
    if (property.valueType === "String" && typeof value !== "string") return "must be a string";
    if (property.valueType === "Boolean" && typeof value !== "boolean") return "must be a boolean";
    if (property.valueType === "Int" && (!numeric || !Number.isInteger(value))) return "must be an integer";
    if (property.valueType === "Float" && !numeric) return "must be a number";
    if (property.valueType === "Enum" && (typeof value !== "string" || !property.constraints?.enumValues?.includes(value))) return `must be one of ${JSON.stringify(property.constraints?.enumValues ?? [])}`;
    if (numeric) { if (property.constraints?.min !== undefined && value < property.constraints.min) return `is below constraint min (${property.constraints.min})`; if (property.constraints?.max !== undefined && value > property.constraints.max) return `exceeds constraint max (${property.constraints.max})`; }
    if (typeof value === "string") { if (property.constraints?.min !== undefined && value.length < property.constraints.min) return `is shorter than min length (${property.constraints.min})`; if (property.constraints?.max !== undefined && value.length > property.constraints.max) return `exceeds max length (${property.constraints.max})`; if (property.constraints?.pattern !== undefined && !new RegExp(property.constraints.pattern).test(value)) return `does not match pattern '${property.constraints.pattern}'`; }
    return undefined;
  }
  private diag(component: string, line: number, code: string, message: string): TamlDiagnostic { return { code, component, line, message, formatted: `[TAML Type Error] Component '${component}' (line ${line}): ${message}` }; }
}

File: crates/taml-core/src/checker.rs
use crate::{
    TamlDiagnostic, TamlValue,
    parser::{Component, Field, Program},
};
use regex::Regex;
use std::collections::{BTreeMap, HashMap, HashSet};
fn valid(field: &Field, value: &TamlValue) -> bool {
    match field.value_type.as_str() {
        "String" => matches!(value, TamlValue::String(_)),
        "Int" => matches!(value, TamlValue::Int(_)),
        "Float" => matches!(value, TamlValue::Int(_) | TamlValue::Float(_)),
        "Boolean" => matches!(value, TamlValue::Boolean(_)),
        "Enum" => matches!(value, TamlValue::String(value) if field.enum_values.contains(value)),
        _ => false,
    }
}
fn number(value: &TamlValue) -> Option<f64> {
    match value {
        TamlValue::Int(value) => Some(*value as f64),
        TamlValue::Float(value) => Some(*value),
        _ => None,
    }
}
pub fn check(program: &Program) -> (Vec<Component>, Vec<TamlDiagnostic>) {
    let mut diagnostics = Vec::new();
    let named: HashMap<&str, &Component> = program
        .components
        .iter()
        .map(|component| (component.name.as_str(), component))
        .collect();
    let mut expanded = Vec::new();
    for component in &program.components {
        let mut values = BTreeMap::new();
        let mut visited = HashSet::new();
        let mut cursor = component.parent.as_deref();
        while let Some(parent_name) = cursor {
            if !visited.insert(parent_name) {
                diagnostics.push(TamlDiagnostic::at(
                    "CYCLIC_INHERITANCE",
                    "cyclic component inheritance",
                    0,
                    1,
                    1,
                ));
                break;
            }
            let Some(parent) = named.get(parent_name) else {
                diagnostics.push(TamlDiagnostic::at(
                    "UNKNOWN_PARENT",
                    format!("unknown parent '{parent_name}'"),
                    0,
                    1,
                    1,
                ));
                break;
            };
            for (key, value) in &parent.values {
                values.entry(key.clone()).or_insert_with(|| value.clone());
            }
            cursor = parent.parent.as_deref();
        }
        values.extend(component.values.clone());
        let mut checked = component.clone();
        checked.values = values;
        if let Some(interface_name) = &checked.interface {
            let Some(interface) = program.interfaces.get(interface_name) else {
                diagnostics.push(TamlDiagnostic::at(
                    "UNKNOWN_INTERFACE",
                    format!("unknown interface '{interface_name}'"),
                    0,
                    1,
                    1,
                ));
                expanded.push(checked);
                continue;
            };
            for (key, field) in &interface.fields {
                if !checked.values.contains_key(key) {
                    if let Some(default) = &field.default {
                        checked
                            .values
                            .insert(key.clone(), (default.clone(), field.line, field.column));
                    } else if !field.optional {
                        diagnostics.push(TamlDiagnostic::at(
                            "MISSING_PROPERTY",
                            format!(
                                "component '{}' is missing required property '{key}'",
                                checked.name
                            ),
                            0,
                            field.line,
                            field.column,
                        ));
                    }
                }
            }
            for (key, (value, line, column)) in &checked.values {
                if let Some(field) = interface.fields.get(key) {
                    if !valid(field, value) {
                        diagnostics.push(TamlDiagnostic::at(
                            "TYPE_ERROR",
                            format!("property '{key}' does not satisfy {}", field.value_type),
                            0,
                            *line,
                            *column,
                        ));
                    }
                    if let Some(number) = number(value) {
                        if field.constraint.min.is_some_and(|min| number < min) {
                            diagnostics.push(TamlDiagnostic::at(
                                "MIN_CONSTRAINT",
                                format!("property '{key}' is below min"),
                                0,
                                *line,
                                *column,
                            ));
                        }
                        if field.constraint.max.is_some_and(|max| number > max) {
                            diagnostics.push(TamlDiagnostic::at(
                                "MAX_CONSTRAINT",
                                format!("property '{key}' exceeds max"),
                                0,
                                *line,
                                *column,
                            ));
                        }
                    }
                    if let (Some(pattern), TamlValue::String(text)) =
                        (&field.constraint.pattern, value)
                    {
                        match Regex::new(pattern) {
                            Ok(expression) if !expression.is_match(text) => {
                                diagnostics.push(TamlDiagnostic::at(
                                    "PATTERN_CONSTRAINT",
                                    format!("property '{key}' does not match pattern '{pattern}'"),
                                    0,
                                    *line,
                                    *column,
                                ))
                            }
                            Err(_) => diagnostics.push(TamlDiagnostic::at(
                                "INVALID_PATTERN",
                                format!("property '{key}' declares an invalid pattern '{pattern}'"),
                                0,
                                field.line,
                                field.column,
                            )),
                            _ => {}
                        }
                    }
                }
            }
        }
        expanded.push(checked);
    }
    (expanded, diagnostics)
}
